blobby/blob.py

```diff
@@ -1,248 +0,0 @@
-        self.seek(0)
-        with open(path, 'wb') as fd_w:
-            fd_w.write(self.read())
-
-    @contextmanager
-    def as_tempfile(self, chunk_size: int = 1000000) -> Iterator[IO[bytes]]:
-        """Context manager which synchronously downloads the object to
-        a temporary file and returns a file descriptor for it.
-
-        Returns: the open temporary file, which will be closed by the
-            context manager.
-        """
-        with tempfile.TemporaryFile('w+b') as fd:
-            chunk = self.read(chunk_size)
-            while chunk:
-                fd.write(chunk)
-                chunk = self.read(chunk_size)
-            fd.seek(0)
-            yield fd
-
-    def seek(self, offset: int, whence: int = io.SEEK_SET) -> int:
-        old_position = self.position
-
-        if whence == io.SEEK_SET:
-            self.position = offset
-        elif whence == io.SEEK_CUR:
-            self.position += offset
-        elif whence == io.SEEK_END:
-            self.position = self.object.content_length + offset
-        else:
-            raise ValueError(
-                f'whence must be one of {self.WHENCE_LIST}, not {whence}.'
-            )
-        if self.position < 0:
-            self.position = old_position
-            raise ValueError(
-                f'seek would result in negative position {self.position}'
-            )
-
-        return self.position
-
-    def seekable(self) -> bool:
-        return True
-
-    def read(self, size: int = -1) -> bytes:
-        # GetObject will be unhappy if we try to read from beyond the
-        # end of the object.
-        if self.position >= len(self):
-            return b''
-
-        if size == -1:
-            range = f'bytes={self.position}-'
-            self.seek(0, io.SEEK_END)
-        else:
-            # But it's fine to specify a range that goes beyond the end
-            # if it starts before.
-            # Subtract one because the range is inclusive.
-            range = f'bytes={self.position}-{self.position + size - 1}'
-            self.seek(size, io.SEEK_CUR)
-
-        # For some reason boto3 type stubs get the type wrong here.
-        # Explanation:
-        #   a GetObject response has a Body parameter which in normal
-        # circumstances is always of type StreamingBody.
-        # Code in botocore.endpoint is responsible for creating the
-        # StreamingBody object. But the result could instead be a string
-        # (if there is an error) or bytes (if handling a different
-        # parameter.)
-        response = self.object.get(Range=range)
-        status = response['ResponseMetadata']['HTTPStatusCode']  # type: ignore
-        if status > 300:
-            raise BlobError(f'Received status code {status} '
-                            f'when getting blob {self.object}')
-        # Make the assumption that the typecheck ignore is making
-        # explicit
-        assert isinstance(response['Body'], StreamingBody), (
```

Comment:
`assert` could be disabled when invoking `python`, since this is quite important I'd personally use an explicit `if not .. raise`. Up to you :)

Reply:
So I'm in two minds about this, and it comes down to the interpretation of an assertion and also what we want to happen in the situation where an unexpected but compatible type is returned: is that an error that we want to *fail* on, or is it just demonstrating that our *assumption is wrong*? @crwr45 since you +1'd do you have an opinion?